ansible/roles/ci/code_style_check/common_resources/files/files_by_extension.py

```diff
@@ -37,29 +39,77 @@ def argparser():
+            full_file_path = os.path.join(dirpath, filename)
+            valid_file = check_for_python_file(full_file_path)
+            if filename in excluded_files:  # Skip excluded files
                 continue
+            if valid_file:
+                output.append(full_file_path)
     return output
 
 
 def gather_all_cpp_files(file_path):
+    """This function walks through the given path collects all c, cpp, h, hpp files It will exclude all python files
+       found in the lint ignore file."""
     output = []
     for dirpath, _, filenames in os.walk(file_path):
+        excluded_files = gather_excluded_files(dirpath, "cpp")  # Gathers all the files to be ignored in this folder
         for filename in filenames:
+            if filename in excluded_files:  # Skip excluded files
+                continue
             if os.path.splitext(filename)[1] in CPP_HEADERS:
-                output.append(f"{dirpath}/{filename}")
+                output.append(os.path.join(dirpath, filename))
     return output
 
+def gather_excluded_files(folder_path, filetype):
+    """Gatheres all of the files to ignore in the lint ignore file (if it exists). A * tells it to skip all files
+       or you can list indiviual files like `exclude_files=test.py,hello.cpp,hi.h`."""
+    folder_files = [f for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))]
+    folder_files_filtered = []
+
+    for f_file in folder_files:
+        if filetype == "python":
+            if check_for_python_file(f_file):
+                folder_files_filtered.append(f_file)
+        else:
+            if os.path.splitext(f_file)[1] in CPP_HEADERS:
+                folder_files_filtered.append(f_file)
+
+    if LINT_IGNORE_FILE in folder_files:  # Check there are actually files to skip
+        re_pattern = re.compile(r'exclude_files=((?:[^,]*,)*[^,]*)')
+        with open(os.path.join(folder_path, LINT_IGNORE_FILE)) as ignore_file:
+            content = ignore_file.readline()
+
+        excluded_files = re_pattern.search(content).group(1)
+        if "*" in excluded_files:  # Tells us to skip all files
+            return folder_files_filtered
+        else:
+            return [f.strip() for f in excluded_files.split(",")]
+
+    return []
+
+def check_for_python_file(file_path):
+    """This function is used to check if a file is a python file. It first checks the extension, if this doesn't return
+       .py then it will check the first line of the file against the python headers. Returns true if the file is a
+       python file."""
+    if os.path.splitext(file_path)[1] == ".py":
+        return True
+
+    try:
+        fline = ""
+        with open(file_path) as python_file:
+            fline = python_file.readline().strip()
+        if any(fline in head for head in PYTHON_HEADERS) and fline != "" and fline != "#":
+            return True
+    except Exception:  # We don't actually care about the exception as it wont fail on python files.
```

Comment:
Why do you use a try-except if that will never fail (arguably)?

Reply:
Ah yes, I remember why I added it, because it opens every file (with utf-8) to check if it has a python header (as we don't always end files in .py) I had to come up with this solution or you get errors like this:


stderr => Traceback (most recent call last):
--
687 | File "/root/.ansible/tmp/ansible-tmp-1672763130.7699652-1473-195906444883797/files_by_extension.py", line 116, in <module>
688 | files = gather_all_python_files(path)
689 | File "/root/.ansible/tmp/ansible-tmp-1672763130.7699652-1473-195906444883797/files_by_extension.py", line 46, in gather_all_python_files
690 | excluded_files = gather_excluded_files(dirpath, "python")  # Gathers all the files to be ignored in this folder
691 | File "/root/.ansible/tmp/ansible-tmp-1672763130.7699652-1473-195906444883797/files_by_extension.py", line 78, in gather_excluded_files
692 | if check_for_python_file(os.path.join(folder_path, f_file)):
693 | File "/root/.ansible/tmp/ansible-tmp-1672763130.7699652-1473-195906444883797/files_by_extension.py", line 106, in check_for_python_file
694 | fline = python_file.readline().strip()
695 | File "/usr/lib/python3.8/codecs.py", line 322, in decode
696 | (result, consumed) = self._buffer_decode(data, self.errors, final)
697 | UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb4 in position 13: invalid start byte

Reply:
So that means that your comment is not exactly true, right? And if you enter this except it might be nice to log something like "Not a Python file, ignoring". Btw, why do you say that not all of our pyhton files end with .py?

Reply:
Potentially, but that also makes it less clear what files get checked, as it goes through literally every item in the repository, even hidden ones (that also causes failures). Because we have quite a few repos that are python code and are structured as such, but does not have a .py extension on the end